modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/UnsortedDistributedCacheQueryReducer.java

```diff
@@ -0,0 +1,211 @@
+        return pageStream.hasNext();
+    }
+
+    /** {@inheritDoc} */
+    @Override public R next() throws IgniteCheckedException {
+        return pageStream.next();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void addPage(@Nullable UUID nodeId, Collection<R> data) {
+        pageStream.addPage(data);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onLastPage() {
+        super.onLastPage();
+
+        loadAllowed = false;
+
+        firstPageLatch.countDown();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void cancel() {
+        Collection<ClusterNode> allNodes = cctx.discovery().allNodes();
+        Collection<ClusterNode> nodes;
+
+        synchronized (queueLock()) {
+            nodes = F.retain(allNodes, true,
+                new P1<ClusterNode>() {
+                    @Override public boolean apply(ClusterNode node) {
+                        return !cctx.localNodeId().equals(node.id()) && subgrid.contains(node.id());
+                    }
+                }
+            );
+
+            rcvd.clear();
+            subgrid.clear();
+        }
+
+        pageRequester.cancelQueryRequest(reqId, nodes, fut.fields());
+
+        pageStream.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void loadPage() {
+        assert !Thread.holdsLock(queueLock());
+
+        Collection<UUID> nodes = null;
+
+        synchronized (queueLock()) {
+            // Loads only queue is empty to avoid memory consumption on additional pages.
+            if (!pageStream.queue.isEmpty())
+                return;
+
+            if (loadAllowed && rcvd.containsAll(subgrid)) {
+                rcvd.clear();
+
+                nodes = new ArrayList<>(subgrid);
+            }
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, false);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void loadAll() throws IgniteInterruptedCheckedException {
+        assert !Thread.holdsLock(queueLock());
+
+        U.await(firstPageLatch);
+
+        Collection<UUID> nodes = null;
+
+        synchronized (queueLock()) {
+            if (loadAllowed && !subgrid.isEmpty())
+                nodes = new ArrayList<>(subgrid);
+        }
+
+        if (nodes != null)
+            pageRequester.requestPages(reqId, fut, nodes, true);
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onPage(@Nullable UUID nodeId, boolean last) {
+        assert Thread.holdsLock(queueLock());
+
+        if (nodeId == null)
+            nodeId = cctx.localNodeId();
+
+        rcvd.add(nodeId);
+
+        if (!loadAllowed && rcvd.containsAll(subgrid) ) {
+            firstPageLatch.countDown();
+            loadAllowed = true;
+        }
+
+        return last && subgrid.remove(nodeId) && subgrid.isEmpty();
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean onNodeLeft(UUID nodeId) {
+        synchronized (queueLock()) {
+            return subgrid.contains(nodeId);
```

Comment:
Here we know the node was left.
Is there any reason to return smth to higher-level instead of releasing a latch instantly?
It is non-obvious if onPage will ever be called.

Reply:
You're correct, it fut.onPage(error) can be invoked in this method. But it looks like as spaghetti. Stack of calls is:
fut.onNodeLeft -> rdc.onNodeLeft -> fut.onPage(error) -> rdc.onPage. 

It's not too much, but looks strange. 

> It is non-obvious if onPage will ever be called.

Why do you think? JavaDocs clearly covers return value of rdc.onNodeLeft.

Reply:
You can have 2 separate calls instead.
fut.onNodeLeft -> rdc.onNodeLeft 
fut.onNodeLeft -> rdc.isFailed/isFinished

Reply:
Renamed method, now it just checks that specified node is running query.

> Is there any reason to return smth to higher-level instead of releasing a latch instantly?

I tried to avoid this, but found that we must release a latch (I mean CountDownLatch) after we mark future as done. It's required for correct return exception to a user in case it's hanging on `awaitInitialization()` method.